jupyterhub/auth.py

```diff
@@ -144,6 +144,98 @@ def _deprecated_db(self):
         """
     ).tag(config=True)
 
+    allow_all = Bool(
+        # dynamic default computed from allowed_users
+        config=True,
+        help="""
+        Any users who can successfully authenticate are allowed to access JupyterHub.
+
+        For backward-compatibility, this is True by default if `allowed_users` is unspecified,
+        False if `allowed_users` is specified.
+
+        Authenticator subclasses may override the default with e.g.::
+
+            @default("allow_all")
+            def _default_allow_all(self):
+                return False
+
+            # or
+
+            @default("allow_all")
+            def _default_allow_all(self):
+                if self.allowed_users or self.allowed_groups:
+                    return False
+                else:
+                    return True
+
+        False is a good idea when the group of users who can authenticate is typically larger
+        than the group users who should have access (e.g. OAuth providers).
+
+        Authenticator subclasses that define additional sources of `allow` config
+        beyond `allowed_users` should specify a default value for allow_all,
+        either always False or `if not any allow_config`.
+
+        This is checked inside `check_allowed`, so subclasses that override `check_allowed`
+        must explicitly check `allow_all` for it to have any effect.
+        This is for safety, to ensure that no Authenticator subclass gets unexpected behavior from `allow_all`.
+
+        .. versionadded:: 5.0
+        """,
+    )
+
+    @default("allow_all")
+    def _default_allow_all(self):
+        if self.allowed_users:
+            return False
+        else:
+            return True
```

Comment:
We now have a dynamic default for `allow_existing_users` and `allow_all`.

Both `allow_all` and `allow_existing_users` in the Authenticator class get their default based on if `allowed_users` is configured truthy or not.

If someone would configure only `allow_existing_users`, they would still get `allow_all` set to True by default, making `allow_existing_users` pointless as everyone is allowed anyhow.

I'm not sure how to handle that, but its certainly very messy to have dynamic defaults. On the other hand, if we fix the values we end up needing to think about how both existing and new authenticators are influenced when using new/old versions of jupyterhub - which is very complicated as well.

I think it may be safe to configure `allow_all` to True `if not self.allowed_users and not self.allow_existing_users` - what do you think?

Reply:
> I think it may be safe to configure allow_all to True if not self.allowed_users and not self.allow_existing_users - what do you think?

I think that's true, good catch. The general rule should be "True if no other allow config", and I totally forgot that `allow_existing_users` _on its own_ could be used to _only_ admit users already in the db, so it definitely qualifies.

WDYT about the case for allow_all=False after this? I think if we set allow_all=False unconditionally, a lot gets simpler because then JupyterHub can be responsible for interpreting `allow_all`, and Authenticators don't need to know about it at all. The only reason we don't do that is to avoid the dynamic default allowing users access. The only downside I see is it's the same breaking change we made in oauthenticator (which was a bit painful), and it makes less sense to prevent that for most other Authenticators. We can still set it to true by default for Dummy (and maybe PAM).

Another mechanism for the dynamic default is to use traitlets metadata to tag all config options that explciitly admit users (e.g. `allowed_users = Set(...).tag(allows_access=True)`), and then allow_all can be False if any of those traits are set. We wuld still have the problem of Authenticators needing to be updated to support this.

Reply:
> WDYT about the case for allow_all=False after this?

I'm very happy about the idea of getting to a state of no longer providing dynamic defaults, but I also find reviewing the consequences of it very complicated. I figure we need to consider the combinations of jupyterhub before/after the change and authenticator classes before/after adjusting to such change.

Reply:
OK, I'll take a stab at simple False and update the docs to match to hopefully cover all of that.

Reply:
OK! The aim is for this to be shipped with 5.x only now right?

Reply:
Yes, definitely. Breaking changes here are okay, as long as they actually improve things for users and don't _expand_ access for any existing config.